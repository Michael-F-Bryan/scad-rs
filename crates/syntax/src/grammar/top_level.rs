use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
