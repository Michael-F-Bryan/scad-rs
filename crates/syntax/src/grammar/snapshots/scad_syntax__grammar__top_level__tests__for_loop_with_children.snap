---
source: crates/syntax/src/grammar/top_level.rs
description: "statement(\n            for(i = [0: 10: 2]) {\n                cube(i);\n                translate(i);\n            })"
---
Node(
    FOR_STATEMENT@0..103
      WHITESPACE@0..13 "\n            "
      FOR_KW@13..16 "for"
      L_PAREN@16..17 "("
      ASSIGNMENTS@17..31
        ASSIGNMENT@17..31
          IDENT@17..18 "i"
          WHITESPACE@18..19 " "
          EQUALS@19..20 "="
          RANGE_EXPR_FROM_TO_STEP@20..31
            WHITESPACE@20..21 " "
            L_BRACKET@21..22 "["
            INTEGER@22..23
              INTEGER@22..23 "0"
            COLON@23..24 ":"
            INTEGER@24..27
              WHITESPACE@24..25 " "
              INTEGER@25..27 "10"
            COLON@27..28 ":"
            INTEGER@28..30
              WHITESPACE@28..29 " "
              INTEGER@29..30 "2"
            R_BRACKET@30..31 "]"
      R_PAREN@31..32 ")"
      BRACED_ACTIONS@32..103
        WHITESPACE@32..33 " "
        L_CURLY@33..34 "{"
        MODULE_INSTANTIATION@34..59
          WHITESPACE@34..51 "\n                "
          IDENT@51..55 "cube"
          L_PAREN@55..56 "("
          ARGUMENTS@56..57
            LOOKUP_EXPR@56..57
              IDENT@56..57 "i"
          R_PAREN@57..58 ")"
          SEMICOLON@58..59
            SEMICOLON@58..59 ";"
        MODULE_INSTANTIATION@59..89
          WHITESPACE@59..76 "\n                "
          IDENT@76..85 "translate"
          L_PAREN@85..86 "("
          ARGUMENTS@86..87
            LOOKUP_EXPR@86..87
              IDENT@86..87 "i"
          R_PAREN@87..88 ")"
          SEMICOLON@88..89
            SEMICOLON@88..89 ";"
        WHITESPACE@89..102 "\n            "
        R_CURLY@102..103 "}"
    ,
)
